### ISSUE TITLE
speed of autocompletion

### ISSUE THREAD

**Issue by dahtah** [gcv/julia-snail#33]
Autocompletion feels very slow on my machine, there's a noticeable lag between pressing Tab and seeing something pop up (this is local, not over a network connection). This means e.g. company-capf is not much use as a completion method because of the lag. I've tried profiling things a bit and so far I've noticed the following: 
- JuliaSnail.lsnames is about 10 times slower than Base.names, but a call to JuliaSnail.lsnames(Base) takes about 20ms so it's unlikely to be the main culprit
- I'm not very familiar with the elisp profiler, but from what I can tell a good chunk of the time is spent in julia-snail--server-response-filter (hard to tell beyond that). 
To what extent can we speed this up? And if we can't gain much, perhaps we should be caching completion lists, and update them once a minute or so?

**Comment by gcv** [comment]
I started looking into this a couple of weeks ago, and found at least two problems.

One, the `lsnames` network call happens more than once. I don't know if company-capf or `completion-at-point` should be blamed for this.

Two, `lsnames` often returns many candidates, which slows down the `read` on the Elisp side. I've found completion to vary from pretty quick on small modules to horrible on large ones. Since many candidates are from `Base` and not related to anything imported into the REPL by the user, they should be cached on the Elisp side and not re-requested (i.e., all the Julia built-ins can safely be requested once). However, we should avoid caching things the user is likely to be working on.

If you watch the Snail process buffer, you'll see what I mean: huge data structures returned multiple times. It definitely needs to get fixed.

**Comment by dahtah** [comment]
As a possible work-around for now, removing julia-snail-completion-at-point from the completions back-ends and using only julia-snail-repl-completion-at-point leads to a quite noticeable speed-up (but still not quite zippy, with occasional freezing that may be due to the parser). What are the completions the former provides that the latter doesn't? Can it be run conditionally?

**Comment by orialb** [comment]
> What are the completions the former provides that the latter doesn't?

The `lsnames` based completion provides fuzzy completion, unlike the `REPLCompletions` based one, not sure if it provides other things that the repl based completions won't provide.
One thing we could play with is using https://github.com/JunoLab/FuzzyCompletions.jl as the completion provider instead of `REPLCompletions` (I think trying it out should be as simple as replacing one function call in `JuliaSnail.replcompletions` but not sure).

I've also noticed that completion can be slow when I work inside a large package (also locally), even when the `lsnames` based completion is disabled (right now I work with `lsnames` completion permanently disabled).

**Comment by dahtah** [comment]
I've tried playing with FuzzyCompletions.jl but it's not quite as straightforward as replacing that one function call in JuliaSnail.replcompletions. What happens if you do that is that company-capf removes all candidates that are not strict prefix matches, at least by default. You end up with the exact same candidates as you have using regular REPL autocompletion. A work-around is to use the company-fuzzy package which has its own fuzzy-matching mechanism and allows inexact matches. Maybe there's a simpler way but I don't know it.
Also, FuzzyCompletions.jl may send thousands of candidates which is going to make the speed problems worse unless we limit it to the top-N matches.

**Comment by gcv** [comment]
Bad news. I did some more careful profiling with REPL-completion only, and found that the filter only takes like 75 microseconds. The culprit is the Parsec-based parser, which we must run to determine module context. That explains why larger source files slow down autocompletion more than small ones. 😕

If there were a way to call Julia's own parser and get back an AST in a form we could work with to extract essential information, it would help; we could replace Snail's parser and solve _a lot_ of problems. I vaguely remember some GitHub tickets about exposing the parser, but do not know if that ever came to fruition or if it's possible now through some APIs or whatever. Maybe one of you could look into it?

That said, calling `read` and `eval` in the filter on 1200+ element lists undoubtedly doesn't help. I'm open to turning off the `lsnames` completion mechanism, since the REPL one works pretty well and, as @dahtah pointed out, `company-capf` doesn't support fuzzy matching anyway.

**Comment by dahtah** [comment]
After investigating this for a bit, here's what I could gather. A first question is whether full parsing is needed for autocompletion, or if we could get away with a few regexps. A second is whether we can replace the parser in julia-snail with Julia's own. 

1. Right now the role of the parser is to figure out the current module context. It seems to me that's overkill, and it has the downside of not working when the syntax is (temporarily) wrong. That happens to me when I'm missing an "end" statement somewhere for a loop: autocompletion errors, which is a bit annoying. It's hard to keep valid syntax all the time. I took a look at how emacs-ess does it for R autocompletion, and it seems to be all regexps, no parsing required except to get symbol at point. 
Maybe the following mechanism is enough: when activating julia-snail inside a Julia package, and every time a .jl file is loaded, figure out where it fits in the module hierarchy (if it is included in a different .jl file inside module M, inherit that module prefix). This way, every buffer has its own julia-snail--module-prefix. When autocompleting, figure out current module by searching backward for unbalanced "module XYZ" statements recursively and appending to julia-snail--module-prefix. This should be much faster than a full parse. 
2. There's some call for more clever parsing to make autocompletion more context-sensitive. To some extent repl-completion already does that by detecting for instance a "\" before point to trigger unicode completion. It'd also be nice to detect a ":: " context to complete types, and a "funcall(; )" context to complete keyword arguments, for instance. I'm not sure whether this calls for a full parser or just some regexp magic. LanguageServer.jl uses a parser, emacs-ess (for R) uses just regexp magic I think. Also, not obvious whether this should be implemented on the Julia or emacs side. 
3. As far as parsers are concerned, there are several options I can see. Julia's own parser is written in a Scheme dialect called femtolisp. Perhaps it'd run in emacs-lisp as well? I'm not enough of a lisper to say. Otherwise, LanguageServer.jl uses After investigating this for a bit, here's what I could gather. A first question is whether full parsing is needed for autocompletion, or if we could get away with a few regexps. A second is whether we can replace the parser in julia-snail with Julia's own. 

1. Right now the role of the parser is to figure out the current module context. It seems to me that's overkill, and it has the downside of not working when the syntax is (temporarily) wrong. That happens to me when I'm missing an "end" statement somewhere for a loop: autocompletion errors, which is a bit annoying. It's hard to keep valid syntax all the time. I took a look at how emacs-ess does it for R autocompletion, and it seems to be all regexps, no parsing required except to get symbol at point. 
Maybe the following mechanism is enough: when activating julia-snail inside a Julia package, and every time a .jl file is loaded, figure out where it fits in the module hierarchy (if it is included in a different .jl file inside module M, inherit that module prefix). This way, every buffer has its own julia-snail--module-prefix. When autocompleting, figure out current module by searching backward for unbalanced "module XYZ" statements recursively and appending to julia-snail--module-prefix. This should be much faster than a full parse. 
2. There's some call for more clever parsing to make autocompletion more context-sensitive. To some extent repl-completion already does that by detecting for instance a "\" before point to trigger unicode completion. It'd also be nice to detect a ":: " context to complete types, and a "funcall(; )" context to complete keyword arguments, for instance. I'm not sure whether this calls for a full parser or just some regexp magic. LanguageServer.jl uses a parser, emacs-ess (for R) uses just regexp magic I think. Also, not obvious whether this should be implemented on the Julia or emacs side. 
3. As far as parsers are concerned, there are several options I can see. Julia's own parser is fast, and is written in a Scheme dialect called femtolisp. Perhaps it would run in emacs-lisp as well? I'm not enough of a lisper to say. Otherwise, LanguageServer.jl uses CSTParser.jl, which seems pretty fast. Having a fast parser opens the possibility of using, e.g., flyspell to highlight syntax errors on the fly.  In another issue tree-sitter was suggested but their Julia parser doesn't seem to be under active development. 
4. To provide useful fuzzy completion using FuzzyCompletions.jl, it would make sense to write a company back-end (that's the approach CIDER takes, for instance). It doesn't look too hard. 
5. Finally, if JuliaSnail.jl is going to depend on non-Base packages, as is likely, it'd make sense to register it as its own package, I think.

**Comment by orialb** [comment]
I was just about to write about the option of [CSTParser.jl](https://github.com/julia-vscode/CSTParser.jl) as the last comment appeared :) 
I'll try to play with it a bit in the context of finding the current module to see if I can get performance improvements for `julia-snail--module-at-point` and the completion. (Even if we end up not doing a full parse for completion as suggested above, maybe CSTParser would be useful for the other parts of Snail).

I also noticed point (1) mentioned above (completion doesn't work if buffer can't be parsed), and I agree it would be great to be able to get completion when some `end` statements are missing (most typical case for me is also when writing a for loop, I usually add the `end` only after writing the `for i in something` part)
I think maybe CSTParser might be still able to partially parse a file even when some end statements are missing (but this is only from a few minutes of playing with it right now, so I might be wrong here).

> It'd also be nice to detect a ":: " context to complete types,
Type completion already works for me if you start typing after `::` (at least the last time I checked). Do you mean that you want it to limit completion suggestions only to types? In the REPL itself it doesn't seem to do that.

**Comment by dahtah** [comment]
>  Type completion already works for me if I start typing after :: (at least the last time I checked). Do you mean that you want it to limit completion suggestions only to types? In the REPL itself it doesn't seem to do that.

Yes, it works, I meant it should be limited to types (not super important, because anything in Julia that starts with a capital letter is supposed to be a type). I'm sure there's plenty of other examples of context-sensitivity I haven't thought about.

**Comment by dahtah** [comment]
Correction: things that start with a capital are either a type or a module

**Comment by orialb** [comment]
>I'm sure there's plenty of other examples of context-sensitivity I haven't thought about.

For me the most useful enhancement would be to have completion of local names inside a function.

I think it would be good to focus on solving the speed issue first because it can severely limit usability when working inside large files (as we have all noticed I guess).

**Comment by orialb** [comment]
In https://github.com/orialb/julia-snail/tree/cstparser you can see a quick experiment to find module at point using CSTParser.jl in order to speed-up completion. 
From my tests it seems to be ~ 10X faster than the current method on a 500 loc file. 
Actually it seems that when using CSTParser to find the module, the majority of the time is just taken by the server request (but I might be wrong here, I didn't test this super carefully).
Probably one could speed things up in this experiment by uniting the server request for finding the module and the completion.

**Comment by dahtah** [comment]
Nice! I had a quick look. A minor issue is that this doesn't yet work on scripts. 
```
type Nothing has no field args
get_module(::Nothing, ::Int64) at JuliaSnail.jl:337
```
I'm a bit worried about this line:
```
return get_module(CSTParser.parse(read(file,String)), point)
```
I don't know how clever the OS is about caching, but won't this trigger a read from the file system at every other keystroke? Also if the user hasn't saved the file yet things will be out of sync. It's probably better to pass buffer contents as a string, I think.

**Comment by orialb** [comment]
>  A minor issue is that this doesn't yet work on scripts.

Thanks for pointing this out. I seem to hit this error you were referring to in a case where `point` is at a position at the end of the file that does not exist in the CST obtained by the parser or something like this.  But I'm quite sure it could be easily solved with a bit more careful implementation of `get_module` if we decide to use this approach.  

> I don't know how clever the OS is about caching, but won't this trigger a read from the file system at every other keystroke? Also if the user hasn't saved the file yet things will be out of sync. It's probably better to pass buffer contents as a string, I think.

Yeah this also a good point, especially with the possibility that the file was not saved (or even does not exist yet). 

(Just to clarify that this is supposed just to be a quick POC for the purpose of this discussion and deciding whether CSTParser should be used or not and definitely things could be improved.)

**Comment by orialb** [comment]
BTW, one problem with sending the buffer text in the server request instead of reading it from file on the server side is that due to the way commands are processed by the server right now (using `Meta.parse`) one has to add some extra backslashes escape double-quotes (\" -> \\\") and dollar signs. Doing this on the elisp side for a long enough file is actually slower than just reading the file on the Julia side. I was using `s-replace`, maybe there is a faster way?.

Also, does someone know how can I replace "$" occurrences in a string with "\$"  ? I only manage to get "\\$" with `(s-replace "$" "\\$" string)`.

**Comment by gcv** [comment]
Catching up on this thread...

I don't see how a regexp based system can work. Once you have nested structures, regexps are no longer an adequate tool. Furthermore, Julia uses the ending delimiter, `end` in its array syntax. Consider doing the matching `end` when the point is inside the `hello` function in this code. This is not a tractable problem for a regexp, which is why I went down the road of writing a parser.

```julia
module Alpha
module Bravo
function hello(x)
  x[10:end]
end
end
end
```

CSTParser.jl is intriguing, but I am reluctant to give Snail Julia dependencies. Right now, it's a fairly simple install of an Emacs package (except vterm, which I'd love to eliminate one day). If Snail becomes a Julia package with dependencies, then we have to start dealing with Pkg interaction _outside_ the user's activated package — so it has to be global, it has to be Julia-version aware, and generally becomes a PITA.

Translating the femtolisp code to Elisp is unlikely to be easy, and I seriously doubt the Elisp interpreter will be fast enough for this type of task anyway.

The built-in femtolisp parser has to be in the Julia binary somewhere. Maybe there's a way to call it from Julia code. I consider calling internal APIs preferable to adding a dependency.

**Comment by orialb** [comment]
>If Snail becomes a Julia package with dependencies, then we have to start dealing with Pkg interaction outside the user's activated package

Couldn't we just provide Project.toml and Manifest.toml files with Snail and then in `JuliaSnail.jl` run `Pkg.activate(@__DIR__)` , load the Snail module , and at the end run `Pkg.activate()` to return to the default environment?  (+ we will need to run Pkg.instantiate at some point ).

I think it would be important to find a Julia base solution for module detection because it felt really smooth using the completion with CSTParser for module detection inside a large file.

**Comment by gcv** [comment]
Into which Julia environment will that install the dependencies? Global?

**Comment by orialb** [comment]
It shouldn't be global, it should install the dependencies in a julia-snail environment (i.e. the environment associated with the folder where `JuliaSnail.jl` is located)

**Comment by orialb** [comment]
But I admit that I am not fully sure whether there could be any side effects when loading code from one environment then switching to another environment and loading code from there (if some dependency is in both environments for example). I could try to investigate this.

**Comment by gcv** [comment]
Please do. I'm worried since the Snail server has to run in a different context from the one the user is developing in.

**Comment by orialb** [comment]
Seems that it could be a problem if you have a package `DEP` as a dependency in two environments `ENV1 , ENV2`with different versions of `DEP` in each environment and you want to load code from both of those environments in the same session. In this case if you first do `using  DEP` inside a module when `ENV1` is activated, and then activate `ENV2` and do `using DEP` it will use the version from `ENV1` as far as I can tell.

I'm not sure whether it is possible to load packages in complete isolation in the same REPL session (i.e. have different modules use different versions of the same package). Still having a Snail environment will make sure that there are no side-effects due to Snail when the user is just using the regular REPL without Snail. 

In case we don't find a good solution without CSTParser, I still think adding CSTParser as a dependency is a worthwhile trade-off.

**Comment by orialb** [comment]
> I'll try to look what is done in Juno (the atom extension) because I think there they also have something running in the same REPL which runs the user code.

I've checked and it seems that Juno installs its dependencies (Atom.jl, Juno.jl) in the global environment. It seems that Juno has many happy users in the Julia community, so I assume that adding the dependencies to the global environment was not causing big problems.
I think we could still avoid changing the global environment by having a Snail environment and pushing it at the end of the `LOAD_PATH`. 

In the specific case we are discussing here,  I think that adding CSTParser could mainly cause some problems in the case where the user wants to develop CSTParser.jl or Tokenize.jl using Snail. But I believe one can find a workaround also for this case.

**Comment by dahtah** [comment]
Catching up on the discussion. Agree with @orialb that a dependency on CSTParser.jl looks worth it. The native parser is accessible using Meta.parse but is less informative and user-friendly, from what I can tell. 

However, if the plan is to rely on a Julia-side parser, that opens the question of how to synchronise buffer state between Julia and emacs. The brute-force approach of sending the entire contents of the buffer every time a reparse is needed may be good enough for a first implementation (it's certainly the simplest option), but it's wasteful and I'm wondering how it will fare on remote setups. The language server approach, AFAICT, is to inform the server (Julia) every time something changes on the client (Emacs). Every insertion/deletion is reflected in the server's internal copy of the document.  In the case of julia-snail this could take the form of e.g. sending just the latest content from buffer-undo-list and implementing a line editor on the Julia side but it sounds like a bit of work.

**Comment by orialb** [comment]
I think one of the complications here is that we want to support a scenario of multiple modules inside one file. 
I must say that this is a scenario that I never encounter myself, for me it is always the case that a file belongs to one module. So it doesn't matter where point is at this file or whether the file has been edited, the mapping file-->module is quite static once I included a file in some module (the only exception is of course when I create the module, but this is usually a one time event).  

How often do you find yourselves having multiple modules in the same file? how important it is to support this scenario?  If we can relax this assumption it might make life easier here.

**Comment by orialb** [comment]
> The native parser is accessible using Meta.parse but is less informative and user-friendly, from what I can tell

We could indeed probably use `Meta.parse` to get the current module information, although as @dahtah  said the output from `CSTParser` is more convenient to work with (for example it already assigns each expression node its position span in the file text).
 
Also note that `CSTParser` seems to be faster, for example for parsing a 412 lines file:
```julia
julia> parsefile(s) = Meta.parse("quote;"*s*";end").args[1]
parsefile (generic function with 1 method)

julia> s = read("codefile.jl");

julia> @btime parsefile(s);
  13.199 ms (1923 allocations: 131.28 KiB)

julia> @btime CSTParser.parse(s,true);
  865.655 μs (7967 allocations: 614.69 KiB)
```

But maybe the biggest advantage of `CSTParser` is that, unlike `Meta.parse`, it will parse the file even when there is some syntax error (e.g. `end` statements missing because they where not typed yet) so Snail features relying on parsing will still be available.

**Comment by gcv** [comment]
Well, this has been a long time coming. I have been (and remain) extremely busy with other things. Also, CSTParser is a pain in the ass to program with and lacks a stable API, so it took quite a bit of head-scratching to make it work right.

Here is a branch with CSTParser fully replacing the original Parsec-based parser. https://github.com/gcv/julia-snail/tree/cstparser

It's based on the ideas on this thread, and on @orialb's branch. It sends (potentially unsaved) buffers to the Julia process by base64-encoding them, which eliminates all escaping problems.

Seems pretty fast to me overall.

This branch also fixes #31 and #5, as well as eliminates #29 from consideration.

@dahtah, @orialb: Can you please test that branch before I merge it into master?

**Comment by dahtah** [comment]
@gcv Thanks for working on this, I'm going to take a look. First issue that cropped up has to do with Pkg.activate(".") on line 477. If you start julia-snail from a non-package directory this will lead to weird behaviour, because the environment in this case should be the default environment and not the local directory (which does not have a Project.toml file). In my case it led to a pretty inscrutable error from initialising Revise.jl from startup.jl. 
A quick fix for now would be to call Pkg.activate(".") iff the local directory contains a Project.toml file.

**Comment by gcv** [comment]
@dahtah: Are you looking at the right branch? There's no automatic `Pkg.activate` anywhere in the code, and [line 477 of `JuliaSnail.jl` is currently empty](https://github.com/gcv/julia-snail/blob/cstparser/JuliaSnail.jl#L477). Snail activates its dependencies using `LOAD_PATH` and references its directory explicitly.

**Comment by dahtah** [comment]
@gcv OK, thanks, my bad, I was looking at an earlier version due to some magit screw-up on my part.

**Comment by gcv** [comment]
Cool. Also, a heads-up: I just pushed a couple of small bugfixes and a precompile performance hack. Please test with the latest changes.

**Comment by orialb** [comment]
@gcv awesome that you went on and incorporated CSTParser, thanks for your work. I think that in addition to the speed improvement, the fact that now one-line function definitions are properly handled is really nice.

I must admit that I am not currently using a lot of Julia for actual work :( , but I will try to play a bit with the CSTParser branch over the weekend. Maybe it will give me a good reason to write some nice Julia code :)

**Comment by dahtah** [comment]
Here's a first problem: start Emacs, open a new file (test.jl), C-c C-z, type a single character in the test.jl buffer ("a"), press Tab after that character. I get an error: (wrong-type-argument characterp nil) and a long stacktrace that seems to point to julia-snail--bslash-before-p.

**Comment by dahtah** [comment]
Here's another one: same procedure, now test.jl contains just the line 
```
a = rand(5)
```
C-c C-k results in an error ```(invalid-function 0.8173980328242143)``` (actual number will vary). I think this one should be relatively easy to fix, let me know if you don't have time, I'll give it a try

**Comment by orialb** [comment]
> Here's a first problem: start Emacs, open a new file (test.jl), C-c C-z, type a single character in the test.jl buffer ("a"), press Tab after that character. I get an error: (wrong-type-argument characterp nil) and a long stacktrace that seems to point to julia-snail--bslash-before-p.

This happens also on master, so unrelated to the parser change. If I recall correctly I introduced this `julia-snail--bslash-before-p` function to allow for completion of latex symbols, and it seems that I didn't properly test its behavior on the first line of a file. I will try to take a look at this when playing around with snail on the weekend.

**Comment by orialb** [comment]
One problem seems to be that one has to instantiate Snail's environment (by calling `Pkg.instantiate()`) in order for CSTParser to be downloaded  on the first time `julia-snail` is called, in case the user never installed CSTParser (I happen to be testing this on a new machine, so I don't have CSTParser installed in any Julia envieronment). I guess instead of doing this check from inside `JuliaSnail.jl` one can just define an elisp function `julia-snail-init` (or something like this) that will take care of it and instruct the user to run it before the first use of snail.

**Comment by gcv** [comment]
Damn it, I forgot to test a fully cold start.

**Comment by orialb** [comment]
I would have never thought about it myself if I hadn't happen to be using a completely new machine, so good thing I got one :)

**Comment by dahtah** [comment]
Here's another problem: sometimes julia-snail loads successfully in the REPL buffer but I get a spurious "no REPL buffer found", I imagine because of a timeout. I suspect load time has increased due to adding a dependency on CSTParser.jl.

**Comment by dahtah** [comment]
And another one, not sure if related to new parser: to reproduce, open a new .jl buffer, C-c C-z, type some stuff, and try to get a completion on "σ2". I get: 
```
StringIndexError("σ2", 2)

string_index_err(::String, ::Int64) at string.jl:12
getindex at string.jl:250 [inlined]
completions(::String, ::Int64, ::Module) at REPLCompletions.jl:573
replcompletion(::String, ::Module) at JuliaSnail.jl:314
top-level scope at none:1
eval at boot.jl:331 [inlined]
eval_in_module(::Array{Symbol,1}, ::Expr) at JuliaSnail.jl:119
macro expansion at JuliaSnail.jl:522 [inlined]
(::Main.JuliaSnail.var"#33#38"{Sockets.TCPSocket})() at task.jl:358
```

**Comment by gcv** [comment]
> Here's a first problem: start Emacs, open a new file (test.jl), C-c C-z, type a single character in the test.jl buffer ("a"), press Tab after that character. I get an error: (wrong-type-argument characterp nil) and a long stacktrace that seems to point to julia-snail--bslash-before-p.

Fixed in 7c76d58.

> C-c C-k results in an error (invalid-function 0.8173980328242143)

Fixed in 164e2ec.

> one has to instantiate Snail's environment (by calling Pkg.instantiate())

Fixed in c295f08. Whoa, that needed a nasty workaround.

> sometimes julia-snail loads successfully in the REPL buffer but I get a spurious "no REPL buffer found"

I bumped up the timeouts in cbbfbc1. Let me know if that helps.

> try to get a completion on "σ2"

Seems to be a bug in Julia's string library, since this blows up the same way without Snail in the picture:

```julia
import REPL.REPLCompletions
REPLCompletions.completions("σ2", 2, Main)
```

Anything else? 😎

**Comment by dahtah** [comment]
Wow, that was quick! Thanks, I will report the REPLCompletions bug elsewhere and test the rest of the changes

**Comment by dahtah** [comment]
The julia people think this is a bug in julia-snail and not REPLCompletions. `REPLCompletions.completions("σ2", 2, Main)` shouldn't be called with index 2. See [here](https://github.com/JuliaLang/julia/issues/38357). Sorry, don't have time to investigate this myself today

**Comment by orialb** [comment]
> The julia people think this is a bug in julia-snail and not REPLCompletions. `REPLCompletions.completions("σ2", 2, Main)` shouldn't be called with index 2. See [here](https://github.com/JuliaLang/julia/issues/38357). Sorry, don't have time to investigate this myself today

I think the problem here  is that [string indices in Julia refer to code units](https://docs.julialang.org/en/v1/manual/strings/#Unicode-and-UTF-8), and not directly to characters. The solution as I understand it would be to make snail call `REPLCompletions.completions(identifier, lastindex(identifier), mod)` instead of what it currently does:
https://github.com/gcv/julia-snail/blob/cbbfbc15fb7c9c855975fefb883a79daf75c1415/JuliaSnail.jl#L333

(in the example of `σ2` we will have `length("σ2")==2` but `lastindex("σ2")==3`).

**Comment by gcv** [comment]
I had no idea strings in Julia behaved like that. Replacing `length` with `lastindex` did the trick. Thanks for doing the research, both of you.

**Comment by gcv** [comment]
Fixed in https://github.com/gcv/julia-snail/commit/18a299dd09eebed6394bd5d832dd1e634d381b72.

**Comment by gcv** [comment]
Unless I hear about other bugs on this branch, I'll merge it into master on Monday.

**Comment by dahtah** [comment]
I haven't found any other bugs, looks pretty solid to me so far

**Comment by gcv** [comment]
Merged! 34498596550b0f81

**Comment by gcv** [comment]
Let's open separate tickets for new bugs on this code.
